certora/specs/MarketInteractions.spec

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+import "ConsistentState.spec";
+
+methods {
+    function _.supply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) external => summarySupply(marketParams, assets, shares, onBehalf, data) expect (uint256, uint256) ALL;
+    function _.withdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) external => summaryWithdraw(marketParams, assets, shares, onBehalf, receiver) expect (uint256, uint256) ALL;
+    function _.idToMarketParams(MetaMorphoHarness.Id id) external => summaryIdToMarketParams(id) expect MetaMorphoHarness.MarketParams ALL;
+
+    function lastIndexWithdraw() external returns(uint256) envfree;
+}
+
+function summaryIdToMarketParams(MetaMorphoHarness.Id id) returns MetaMorphoHarness.MarketParams {
+    MetaMorphoHarness.MarketParams marketParams;
+
+    // Safe require because:
+    // - markets in the supply/withdraw queue have positive lastUpdate (see LastUpdated.spec)
+    // - lastUpdate(id) > 0 => marketParams.id() == id is a verified invariant in Morpho Blue.
+    require Util.libId(marketParams) == id;
+
+    return marketParams;
+}
+
+function summarySupply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns(uint256, uint256) {
+    assert shares == 0;
+    assert onBehalf == currentContract;
+    assert data.length == 0;
+
+    MetaMorphoHarness.Id id = Util.libId(marketParams);
+    // Safe require because it is a verified invariant
+    require hasSupplyCapIsEnabled(id);
+
+    // Check that all markets on which MetaMorpho supplies are enabled markets.
+    assert config_(id).enabled;
+
+    // NONDET summary, which is sound because all non view functions in Morpho Blue are abstracted away.
+    return (_, _);
```

Comment:
what does this do?

Reply:
It's like if it returned a random value. This is the same as saying that you don't know anything about the return value. This is the safest option for returned values, but it omits storage change. I think it's fine to summarize it this way because all other functions of Morpho Blue are summarized as being able to do anything (notably on storage) on contracts other than MM, so the fact that this does not modify the storage should not impact the soundness of the verification

Reply:
>It's like if it returned a random value. This is the same as saying that you don't know anything about the return value. This is the safest option for returned values, 

I get this

>but it omits storage change. I think it's fine to summarize it this way because all other functions of Morpho Blue are summarized as being able to do anything (notably on storage) on contracts other than MM, so the fact that this does not modify the storage should not impact the soundness of the verification

but not this

the problem could be that we don't specify how does this function change the storage of Morpho, and of the other contracts (tokens etc...) right?

but you are saying that it's ok, because the prover usually assumes that MM can change the storage of other contract in any way. But I don't get why it's ok to assume no storage change on MM and Blue (imagine a blue function disables a market on MM).

Reply:
> the problem could be that we don't specify how does this function change the storage of Morpho, and of the other contracts (tokens etc...) right?
but you are saying that it's ok, because the prover usually assumes that MM can change the storage of other contract in any way. 

Yes exactly. The default summarization is called HAVOC_ECF and it allows changes on the storage of other contracts, but leave the current contract  (MM in this case) unchanged, and return any value. The NONDET summary is to leave every contract unchanged, and return any value. So the difference is in the change of storage of other contracts, but as you said, it should not impact the verification since it already can change arbitrarily for every other function calls.

> But I don't get why it's ok to assume no storage change on MM and Blue (imagine a blue function disables a market on MM).

So this is more about the HAVOC_ECF summarization and why it's sound. One good case is when you only target contracts that were deployed previously and don't allow to call arbitrary other contracts. But this is usually not trivial, and relies on the fact that the contracts you call are trusted. In our case we don't call Blue with callbacks and we verified that there are no untrusted calls, so this summarization should work (but you don't know anything about the details of the functions called)